### DISCUSSION TITLE
How to debug HTTP_PROXY config?

### DISCUSSION THREAD

**Discussion by pat-s** [coder/code-server#6784] [Q&A] [answered]
I am deploying `code-server` in an environment with an HTTP proxy in a container. Env vars `HTTP_PROXY` and `HTTPS_PROXY` are set in the container.

However, when installing extensions, I am getting the error

```
ERROR: getaddinfo ENOTFOUND open-vsx.org
```

The nameservers in the container are internal ones and only resolve internal addresses. So far, so good. However, I would expect `code-server` to utilize the proxy for DNS resolution (which is able to resolve public DNS records). 
Yet even when starting `code-server` with `--log debug` I don't see any proof that the env vars are actually honored and picked up during the attempt to install an extension.

Note: I am able to manually download any `.vsix` extension from open-vsx.org in the `code-server` terminal with the env vars set. So the proxy is working in general. What is **not working** is the attempt to install extensions when clicking on the "Install extension XY" in the extension overview. 

I've already tried to set the proxy explicitly in the user settings in `code-server`, same result.

Is there any additional configuration I need to do to tell `code-server` to use the proxy for DNS resolution? Or any way I can ensure the proxy is actually picked up?

**Comment by code-asher** [comment]
Just to make sure I am on the same page:

1. Running `code-server --install-extension <extension id>` in the terminal: works
2. Searching extensions in the sidebar: works
3. Clicking `install` in the sidebar or in the extension details: broken

My first guess is that the environment variables are set in your shell, but not in code-server's environment.  Can you share more details on how you are running code-server and setting environment variables in the container?

I believe you can also set the proxy in the settings, so that might be worth a shot just to see if it is an environment variable issue.

**Reply by pat-s** [reply]
@code-asher Thanks for helping!

(1) I can't test this since I am using a remote launcher approach to launch code-server which does not have `code-server` bundled in the resulting container. I am launching `code-server` through [Posit Workbench](https://docs.posit.co/ide/server-pro/vscode_sessions/vs_code_sessions.html).

(2) Correct

(3) Correct

Note: I am deploying code-server included in Posit Workbench for several years already. The usual way to configure `http_proxy` is to set it in a file called `launcher-env` which is then [populated as the session environment](https://support.posit.co/hc/en-us/articles/16973467360919-Setting-Environment-Variables-in-Posit-Workbench). This usually works fine and so far I never had an issue. I also have other environments which work fine using this approach. 
In addition, I am setting the proxy vars on the user level, which explains why I can manually download extensions through the proxy using `curl`.

Though what leaves me puzzled is that in this scenario, the error message indicated that the proxy is not used for DNS resolution: instead, the local nameserver is used and then code-server fails as this one cannot resolve open-vsx.org. Usually it should go through the proxy which is able to resolve public domains. **However**, I would expect that searching for extensions then also fails. But this works! Shouldn't this be a `code-server` action, in the same way as installing an extension is?

However, I cannot really debug/fully proof that `code-server` is actually using the proxy env var. Even with debug logging enabled, I don't see a log line which confirms the use.

**Reply by code-asher** [reply]
> I can manually download extensions through the proxy using curl.

Ahhhhh with `curl`, not  `code-server` like I was thinking.  Makes sense.

Does the proxy have a self-signed certificate or anything like that?  One common thing I see is that VS Code seems to struggle to read system-wide certificates, so sometimes you have to set the `NODE_EXTRA_CA_CERTS` environment variable.  Although, if this used to work then it seems unlikely to be the problem.

> I would expect that searching for extensions then also fails. But this works! Shouldn't this be a code-server action, in the same way as installing an extension is?

Extension searching happens from the browser, not the backend, so often searching will work even when installation does not.  

> Even with debug logging enabled, I don't see a log line which confirms the use.

There is the `NODE_DEBUG` environment variable though which can be set to `http,https` to make Node output logs about every HTTP request it makes.  I experimented with this command:

```
NODE_EXTRA_CA_CERTS=$HOME/.mitmproxy/mitmproxy-ca-cert.pem NODE_DEBUG=http,https http_proxy=http://localhost:8085 code-server --auth none --disable-telemetry
```

Where I have `mitmproxy` sitting on port 8085 and my code-server version is 4.23.1.

One of the things I see is this (trimmed for brevity):

```
HTTP 8623: createConnection open-vsx.org:443: [Object: null prototype] {
  hostname: 'open-vsx.org',
  port: 443,
  protocol: 'https:',
  method: 'GET',
  agent: HttpsProxyAgent {
    proxy: URL {
      href: 'http://localhost:8085/',
      origin: 'http://localhost:8085',
      protocol: 'http:',
      host: 'localhost:8085',
      hostname: 'localhost',
      port: '8085',
    },
}
```

I was not able to find any other mentions of the proxy either even with `--verbose` (it really should be added somewhere) but this should definitively prove whether the proxy is being used.

**Reply by pat-s** [reply]
Thanks, that is super helpful and gets me one step further.

In my case, I have `agent: null` instead of the `HttpsProxyAgent` as in your example. Hence, likely the proxy is not being used/picked up in my environment.

Now I only need to find out why that is, as the proxy env vars are set in the pod in which `code-server` is running.

**Reply by code-asher** [reply]
Very curious!  They are not being unset in a `.bashrc` or `/etc/profile` or something are they?

Or is `no_proxy` set, maybe?

I was looking into the module VS Code uses to do the proxying (https://github.com/microsoft/vscode-proxy-agent) and noticed a different environment variable `DEBUG` provided by the `debug` package they and their dependencies are using so I tried `DEBUG=*` and it gave me some new lines:

```
2024-05-09T01:16:30.712Z proxy-agent Proxy URL: 'http://localhost:8085'
2024-05-09T01:16:30.712Z https-proxy-agent Creating new HttpsProxyAgent instance: 'http://localhost:8085/'
```

Which seems to correlate to https://github.com/TooTallNate/proxy-agents/blob/70023c12abe0d014004af6309ff7d0fdbaa60875/packages/proxy-agent/src/index.ts#L126

Which itself seems to rely on https://github.com/Rob--W/proxy-from-env/blob/095d1c26902f37a12e22bea1b8c6b67bf13fe8cd/index.js#L24.  Looks like they check a few other things too like `npm_config_http_proxy` so might be worth checking all those out.

**Reply by pat-s** [reply]
> Very curious! They are not being unset in a .bashrc or /etc/profile or something are they?

No.

> Or is no_proxy set, maybe?

No.

> I was looking into the module VS Code uses to do the proxying ([microsoft/vscode-proxy-agent](https://github.com/microsoft/vscode-proxy-agent)) and noticed a different environment variable DEBUG provided by the debug package they and their dependencies are using so I tried DEBUG=* and it gave me some new lines:

Thanks! I've set it but don't see these lines. What I see is that `proxy-agent` is never being set/defined. Neither is `https-proxy-agent`. 
This might in the end explain the issue in the first place. However, I don't understand why they are not set.

The env vars are defined, in any kind of spelling (small case, upper case). I've played around with adding/removing the `http://` prefixes. Whatever I do, they are not picked up by `code-server`.

`NODE_EXTRA_CA_CERTS` is set to point to the proxy cert.

I've now also tried to invoke the installation command locally in a shell with `HTTP_PROXY` being set but the proxy-agent is still not being used. I might be missing something here but I do not know what else I should try 🤔 

This is quite a bummer from a user PoV as I am trying to configure this for a team of engineers and without the extensions working "as they should", we might need to ditch the whole idea.

code-server 4.22.1

<details><summary>Log</summary>
<p>

/usr/lib/rstudio-server/bin/code-server/bin/code-server --install-extension reditorsupport.r
  router:route new '*' +0ms
  router:layer new '*' +0ms
  router:layer new '/' +1ms
  router:route new '*' +4ms
  router:layer new '*' +1ms
  router:route get * +1ms
  router:layer new '/' +0ms
  router:route new '/' +3ms
  router:layer new '/' +3ms
  router:route get / +0ms
  router:layer new '/' +1ms
  router:route new '/' +1ms
  router:layer new '/' +0ms
  router:route get / +0ms
  router:layer new '/' +0ms
  router use '/' <anonymous> +0ms
  router:layer new '/' +21ms
  router:route new '/' +22ms
  router:layer new '/' +1ms
  router:route get / +0ms
  router:layer new '/' +0ms
  router:route new '/' +0ms
  router:layer new '/' +0ms
  router:route post / +0ms
  router:layer new '/' +0ms
  router:route new '/' +1ms
  router:layer new '/' +1ms
  router:route get / +1ms
  router:layer new '/' +1ms
  router:route new '/check' +1ms
  router:layer new '/check' +1ms
  router:route get /check +1ms
  router:layer new '/' +1ms
  router:route get /check +0ms
  router:layer new '/' +0ms
Installing extensions...
HTTP 3801: call onSocket 0 0
HTTP 3801: createConnection open-vsx.org:443::::::::true::::::::::::: [Object: null prototype] {
  hostname: 'open-vsx.org',
  port: 443,
  protocol: 'https:',
  path: null,
  method: 'POST',
  headers: {
    'X-Market-Client-Id': 'VSCode 1.87.2',
    'User-Agent': 'VSCode 1.87.2 (code-server)',
    'X-Market-User-Id': 'a07ce755-5a7b-4267-9843-31577034cb14',
    'VSCode-SessionId': 'someValue.machineId',
    'Content-Type': 'application/json',
    Accept: 'application/json;api-version=3.0-preview.1',
    'Accept-Encoding': 'gzip',
    'Content-Length': '241'
  },
  agent: null,
  rejectUnauthorized: true,
  _defaultAgent: Agent {
    _events: [Object: null prototype] {
      free: [Function (anonymous)],
      newListener: [Function: maybeEnableKeylog]
    },
    _eventsCount: 2,
    _maxListeners: undefined,
    defaultPort: 443,
    protocol: 'https:',
    options: [Object: null prototype] { noDelay: true, path: null },
    requests: [Object: null prototype] {},
    sockets: [Object: null prototype] {
      'open-vsx.org:443::::::::true:::::::::::::': []
    },
    freeSockets: [Object: null prototype] {},
    keepAliveMsecs: 1000,
    keepAlive: false,
    maxSockets: Infinity,
    maxFreeSockets: 256,
    scheduling: 'lifo',
    maxTotalSockets: Infinity,
    totalSocketCount: 0,
    maxCachedSessions: 100,
    _sessionCache: { map: {}, list: [] },
    [Symbol(kCapture)]: false
  },
  host: 'open-vsx.org',
  noDelay: true,
  servername: 'open-vsx.org',
  _agentKey: 'open-vsx.org:443::::::::true:::::::::::::'
}
HTTPS 3801: createConnection [Object: null prototype] {
  hostname: 'open-vsx.org',
  port: 443,
  protocol: 'https:',
  path: null,
  method: 'POST',
  headers: {
    'X-Market-Client-Id': 'VSCode 1.87.2',
    'User-Agent': 'VSCode 1.87.2 (code-server)',
    'X-Market-User-Id': 'a07ce755-5a7b-4267-9843-31577034cb14',
    'VSCode-SessionId': 'someValue.machineId',
    'Content-Type': 'application/json',
    Accept: 'application/json;api-version=3.0-preview.1',
    'Accept-Encoding': 'gzip',
    'Content-Length': '241'
  },
  agent: null,
  rejectUnauthorized: true,
  _defaultAgent: Agent {
    _events: [Object: null prototype] {
      free: [Function (anonymous)],
      newListener: [Function: maybeEnableKeylog]
    },
    _eventsCount: 2,
    _maxListeners: undefined,
    defaultPort: 443,
    protocol: 'https:',
    options: [Object: null prototype] { noDelay: true, path: null },
    requests: [Object: null prototype] {},
    sockets: [Object: null prototype] {
      'open-vsx.org:443::::::::true:::::::::::::': []
    },
    freeSockets: [Object: null prototype] {},
    keepAliveMsecs: 1000,
    keepAlive: false,
    maxSockets: Infinity,
    maxFreeSockets: 256,
    scheduling: 'lifo',
    maxTotalSockets: Infinity,
    totalSocketCount: 0,
    maxCachedSessions: 100,
    _sessionCache: { map: {}, list: [] },
    [Symbol(kCapture)]: false
  },
  host: 'open-vsx.org',
  noDelay: true,
  servername: 'open-vsx.org',
  _agentKey: 'open-vsx.org:443::::::::true:::::::::::::',
  encoding: null
}
HTTP 3801: sockets open-vsx.org:443::::::::true::::::::::::: 1 1
HTTP 3801: write ret = true
HTTP 3801: outgoing message end.
(node:3801) Warning: Setting the NODE_DEBUG environment variable to 'http' can expose sensitive data (such as passwords, tokens and authentication headers) in the resulting log.
(Use `node --trace-warnings ...` to show where the warning was created)
HTTP 3801: SOCKET ERROR: getaddrinfo ENOTFOUND open-vsx.org Error: getaddrinfo ENOTFOUND open-vsx.org
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:108:26)
Error while installing extensions: getaddrinfo ENOTFOUND open-vsx.org
getaddrinfo ENOTFOUND open-vsx.org

</p>
</details>

**Reply by code-asher** [reply]
I tried 4.22.1 just in case it has a bug but that worked for me as well.  Here is what I got (minus the router debug lines).  I used a nonexistent domain this time, just wanted to make sure it gets set even in that case.

```
coder@dev27:~/coder$ printenv | grep -i proxy
VSCODE_PROXY_URI=https://{{port}}--dev--dev27--code-asher--apps.dev.coder.com
coder@dev27:~/coder$ code-server --version
4.22.1 760d1318e98945d05133c6121f99541ca7a39bf8 with Code 1.87.2
coder@dev27:~/coder$ http_proxy=http://does-not-exist DEBUG=* code-server --install-extension vscodevim.vim
Installing extensions...
  https-proxy-agent Creating new HttpsProxyAgent instance: 'http://does-not-exist/' +0ms
  https-proxy-agent Creating `net.Socket`: { ALPNProtocols: [ 'http/1.1' ], host: 'does-not-exist', port: 80, auth: null, rejectUnauthorized: true, keepAlive: true } +5ms                                                  
  https-proxy-agent:parse-proxy-response onerror Error: getaddrinfo ENOTFOUND does-not-exist     at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:108:26) { errno: -3008, code: 'ENOTFOUND', syscall: 'getaddrinfo', hostname: 'does-not-exist' } +0ms                                                                             
Error while installing extensions: getaddrinfo ENOTFOUND does-not-exist
getaddrinfo ENOTFOUND does-not-exist
```

Sorry this is being so difficult, I wish I knew the solution.  If I had a way to reproduce I would step through the code and find the problem spot.

**Reply by pat-s** [reply]
Thanks for the help!

I digged a bit more and here's what I found:

It is likely a DNS resolution issue in the first place. I am not yet sure what `code-server`'s part in that one is.

-> The local nameservers in that specific environment do not resolve public domains. If I declare them explicitly in `/etc/hosts` (besides more intermediate microsoft.xxx.yyy domains), I get over this step and finally get to a point where clicking on the "install" button does not immediately error but just runs infinite (yet this is nothing more than a debug solution as code-server is running in a k8s pod).

Inspecting the logs then shows some TLS issue. This is strange again as I have set `NODE_EXTRA_CA_CERTS` to the bundle containing all certs (this bundle is working for many other applications).

So right now I see it as follows:

- `code-server` is not directly honoring the proxy for DNS resolution, i.e. it first uses the local nameservers and does not directly go through the proxy
- There's an issue with the TLS certs which I am not sure yet which side is "to blame". This bundle is working fine with other apps and AFAIK, the only thing needed is to set `NODE_EXTRA_CA_CERTS` to the CA cert bundle for code-server to honor these?

For now, I don't know what to next. The people responsible for the DNS tell me it is an application issue and the local nameservers will never resolve public domains. The application should directly talk to the HTTP_PROXY for DNS resolution.
Their claim is backed by the fact that other applications running in this environment do so when they have to resolve public domains. 

For now, I have to tell users that they need to install every extension via manual downloads + install. Which is of course a pain, the same goes for updating. Hence, I am still interested in a long-term solution to get this working, otherwise it is likely that `code-server` will have to be removed again.

**Reply by code-asher** [reply]
That is very interesting.  :thinking:  Hmm so maybe I can reproduce if I have a DNS that is only accessible through the proxy.  Honestly I am not too sure how to set that up; maybe I can do something with iptables.

> the only thing needed is to set NODE_EXTRA_CA_CERTS to the CA cert bundle for code-server to honor these?

I believe this is exactly right, although maybe there is some caveat I am not aware of.

I looked around the proxy agent for issues related to DNS but did not find anything.  https://github.com/search?q=repo%3ATooTallNate%2Fproxy-agents+dns&type=issues

Maybe if we can come up with a simple Node repro we can post there.  Oh they have an example actually:

```javascript
import * as https from 'https';
import { ProxyAgent } from 'proxy-agent';

// The correct proxy `Agent` implementation to use will be determined
// via the `http_proxy` / `https_proxy` / `no_proxy` / etc. env vars
const agent = new ProxyAgent();

// The rest works just like any other normal HTTP request
https.get('https://jsonip.com', { agent }, (res) => {
  console.log(res.statusCode, res.headers);
  res.pipe(process.stdout);
});
```

I used the Node bundled with my copy of code-server:

```shell
$ mkdir test
$ vim test.mjs # paste contents above
$ npm install proxy-agent
$ NODE_EXTRA_CA_CERTS=$HOME/.mitmproxy/mitmproxy-ca-cert.pem DEBUG=* https_proxy=http://localhost:8085 ~/.local/lib/code-server-4.22.1/lib/node test.mjs
```

One difference I noticed is that I have to use `https_proxy` where with code-server I could get by with only `http_proxy`, but I think this is because VS Code essentially sets `https_proxy` equal to `http_proxy` if not set, so probably not related to the issue.

Hopefully it reproduces here, but if not the problem must be in code-server or VS Code...

--------------------

Another thing we might try is downloading the official VS Code, and then using `code --install-extension` to see if it has the same issues.  That could narrow down whether code-server specifically has messed something up in the way it wraps VS Code or if it is in VS Code itself.

If it is in VS Code itself, we could try opening an issue in https://github.com/microsoft/vscode where they probably understand the inner workings a lot better than I do.  If not, then, at least it narrows things down.

It is also possible to use their web version directly without the code-server wrapper by running `code --serve-web` with the official VS Code; note this will download their web server though, it is not bundled with the VS Code distribution.

**Reply by pat-s** [reply, marked as answer]
Good news! I got is solved.

Turns out it was a wrong specification of `NODE_EXTRA_CA_CERTS`! 🤯️ 
Background: I assumed that `/etc/ssl/certs/ca-bundle.crt` is automatically read, also based on [your comment here](https://github.com/coder/code-server/issues/6051#issuecomment-1457206762). And that `NODE_EXTRA_CA_CERTS` only adds *additional* certs.

Turns out that in this case, setting `NODE_EXTRA_CA_CERTS=/etc/ssl/certs/ca-bundle.crt` solved the issue directly as all required certs were in there and no important ones were in the bundle which I initially referenced via `NODE_EXTRA_CA_CERTS`.

Take-away: maybe it is useful to add that piece of information to the docs somewhere?

Overall: thanks for your great patience and tremendous help here: In the end the one-liner to efficiently debug the issue lead me to find the issue and made it easier to swap different CA bundles.

**Reply by code-asher** [reply]
Ahhhhhhh very interesting!  I believe VS Code meant to read certificates installed on those system paths but I bet there is a bug there; I made a note to dig into it later.  Sorry that ended up being misleading!  Definitely agree this needs to be documented.

Really glad you figured it out!!
